### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by aliabbani**
## Blog App - processing data in models
In this milestone I:
- Create model classes for all entities as shown in the ERD diagram.
- Set up associations between models.
- Create a method in the user model that returns the 3 most recent posts for a given user.
- Create a method in the post model that updates the posts counter for a user.
- Create a method in the post model which returns the 5 most recent comments for a given post.
-  Create a method in the comment model that updates the comments counter for a post.
-  Create a method in the like model that updates the likes counter for a post.
- Create a method called current_user in the ApplicationController that returns the first user from a database. Without worrying about proper authentication.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
+end
+
+# A method that returns the 3 most recent posts for a given user.
+def recent_3_posts
+  post.last(3)
+end
```
Comment (DeliceLydia) [inline app/models/user.rb]: - [x] Please setup associations between models like for user model you should add association between user, `posts`, `comments`, `likes`

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+end
+
+# if we add or delete a post from a user
+# the posts_counter will be updated in the user table
+
+# A method which returns the 5 most recent comments for a given post.
+def recent_5_comments
+  # Comments.last(5)
+  Comments.select('text').joins(:posts).last(5)
+end
```
Comment (DeliceLydia) [inline app/models/post.rb]: - [x] Please add associations between post and other models like `comments` and `likes`

Comment (DeliceLydia) [inline app/models/post.rb]: - [x] These 2 methods should be inside the `Post` class
- [x] I suggest that you use [rails active callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) `after_save` to call the `update_post_counter` method every time a new post has been saved to the database, then make the `update_post_counter` method private

File: app/models/user.rb (hunk 1 above)
Comment (DeliceLydia) [inline app/models/user.rb]: - [x] This `recent_3_posts` method should be inside `User` class

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+end
+
+# A method that updates the likes counter for a post.
+def update_likes_counter
+  Post.select.update(:likes_counter).count
+end
```
Comment (DeliceLydia) [inline app/models/like.rb]: - [x] The `update_likes_counter` should be inside `Like` class
- [x] I suggest that you use [rails active callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) `after_save` to call the `#update_likes_counter` method every time a new like has been saved to database, then make the `update_likes_counter` method private

File: app/models/comment.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+end
+
+# A method that updates the comments counter for a post.
+def update_comments_counter
+  Post.select.update(:comments_counter).count
+end
```
Comment (DeliceLydia) [inline app/models/comment.rb]: - [x] `update_comments_counter` method should be inside `Comment` class
- [x] I suggest that you use [rails active callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) `after_save` to call the `#update_comments_counter` method every time a new like has been saved to database, then make the `update_comments_counter` method private

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,2 +1,10 @@
 class ApplicationController < ActionController::Base
 end
+
+# Create a method called current_user in your ApplicationController.
+# A current_user method should return the first user from a database.
+# Don't worry about proper authentication at this time. You will add it in the upcoming projects.
+
+def current_user
+  User.first
+end
```
Comment (DeliceLydia) [inline app/controllers/application_controller.rb]: - [x] I suggest that you make this `current_user` method a class method

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
## Hi @aliabbani, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there! major highlights for your project:
- Kudos to the awesome PR
- Followed git workflow
- No linter errors


## Required Changes ♻️
_Check the comments under the review._



### Cheers and Happy coding!👏👏👏

Feel free to leave a question, or a comment on the PR thread if something is not clear 100%. 

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (DeliceLydia) [reply app/models/post.rb]: - [x] Remember to remove unnecessary commented codes

File: app/models/user.rb (hunk 1 above)
Comment (aliabbani) [reply app/models/user.rb]: Done.

File: app/models/post.rb (hunk 2 above)
Comment (aliabbani) [reply app/models/post.rb]: Done.

Comment (aliabbani) [reply app/models/post.rb]: Done.

File: app/models/user.rb (hunk 1 above)
Comment (aliabbani) [reply app/models/user.rb]: Done.

File: app/models/like.rb (hunk 3 above)
Comment (aliabbani) [reply app/models/like.rb]: Done.

File: app/models/comment.rb (hunk 4 above)
Comment (aliabbani) [reply app/models/comment.rb]: Done.

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (aliabbani) [reply app/controllers/application_controller.rb]: Done.

**Review by arslanbisharat — APPROVED** [review]
# Approved 👏 

Hi there, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

To Highlight: 
- The Pull Request has a proper title and description. ✔️ 
- You have added a descriptive Readme file. Good Job. ✔️ 
- There is no linter error is present. ✔️ 
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️ 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of code reviews per project which you can see from your dashboard. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)_

**Pull request merged**